File: reports/general/uncatcats.py
# Copyright 2008, 2013 bjweeks, MZMcBride, Tim Landscheidt

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
Report class for uncategorized categories
"""

import reports

class report(reports.report):
    def get_title(self):
        return 'Uncategorized categories'

    def get_preamble_template(self):
        return 'Uncategorized categories; data as of <onlyinclude>%s</onlyinclude>.'

    def get_table_columns(self):
        return ['Category', 'Length', 'Members', 'Last edit', 'Last user']

    def get_table_rows(self, conn):
        cursor = conn.cursor()
        cursor.execute('''
        /* uncatcats.py SLOW_OK */
        SELECT
          page_title,
          page_len,
          cat_pages,
          rev_timestamp,
          CONVERT(rev_user_text USING utf8)
        FROM revision
        JOIN
        (SELECT
           page_id,
           CONVERT(page_title USING utf8) AS page_title,
           page_len,
           cat_pages
         FROM category
         RIGHT JOIN page
         ON cat_title = page_title
         LEFT JOIN categorylinks
         ON page_id = cl_from
         WHERE cl_from IS NULL
         AND page_namespace = 14
         AND page_is_redirect = 0) AS pagetmp
        ON rev_page = pagetmp.page_id
        AND rev_timestamp = (SELECT
                               MAX(rev_timestamp)
                             FROM revision AS last
                             WHERE last.rev_page = pagetmp.page_id);
        ''')

        for page_title, page_len, cat_pages, rev_timestamp, rev_user_text in cursor:
            if page_title:
                page_title = u'{{clh|1=%s}}' % page_title
            if not cat_pages:
                cat_pages = ''
            yield [page_title, str(page_len), str(cat_pages), rev_timestamp, rev_user_text]

        cursor.close()

File: dbreps2/src/general/uncatcats.rs
/*
Copyright 2008 bjweeks, MZMcBride
Copyright 2021 Kunal Mehta <legoktm@debian.org>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use anyhow::Result;
use dbreps2::{Frequency, Report};
use mysql_async::prelude::*;
use mysql_async::Conn;

pub struct Row {
    page_title: String,
    page_len: u32,
    cat_pages: u32,
    rev_timestamp: String,
    actor_name: String,
}

pub struct UncatCats {}

#[async_trait::async_trait]
impl Report<Row> for UncatCats {
    fn title(&self) -> &'static str {
        "Project:Database reports/Uncategorized categories"
    }

    fn frequency(&self) -> Frequency {
        Frequency::Weekly
    }

    fn query(&self) -> &'static str {
        r#"
/* uncatcats.rs SLOW_OK */
SELECT
  page_title,
  page_len,
  cat_pages,
  rev_timestamp,
  actor_name
FROM
  revision
  JOIN actor ON rev_actor = actor_id
  JOIN (
    SELECT
      page_id,
      page_title,
      page_len,
      cat_pages
    FROM
      category
      RIGHT JOIN page ON cat_title = page_title
      LEFT JOIN categorylinks ON page_id = cl_from
    WHERE
      cl_from IS NULL
      AND page_namespace = 14
      AND page_is_redirect = 0
  ) AS pagetmp ON rev_page = pagetmp.page_id
  AND rev_timestamp = (
    SELECT
      MAX(rev_timestamp)
    FROM
      revision AS last
    WHERE
      last.rev_page = pagetmp.page_id
  );
"#
    }

    async fn run_query(&self, conn: &mut Conn) -> Result<Vec<Row>> {
        let rows = conn
            .query_map(
                self.query(),
                |(
                    page_title,
                    page_len,
                    cat_pages,
                    rev_timestamp,
                    actor_name,
                )| Row {
                    page_title,
                    page_len,
                    cat_pages,
                    rev_timestamp,
                    actor_name,
                },
            )
            .await?;
        Ok(rows)
    }

    fn intro(&self) -> &'static str {
        "Uncategorized categories; {asof}."
    }

    fn headings(&self) -> Vec<&'static str> {
        vec!["Category", "Length", "Members", "Last edit", "Last user"]
    }

    fn format_row(&self, row: &Row) -> Vec<String> {
        // TODO: Improve this interface
        let mut fmt = vec![];
        fmt.push(format!("{{{{clh|1={}}}}}", &row.page_title));
        fmt.push(row.page_len.to_string());
        fmt.push(row.cat_pages.to_string());
        fmt.push(row.rev_timestamp.to_string());
        fmt.push(row.actor_name.to_string());

        fmt
    }
}
